backend/dataline/services/llm_flow/utils.py

```diff
@@ -0,0 +1,140 @@
+from typing import Any, Iterable, Self, cast
+from langchain_community.utilities.sql_database import SQLDatabase
+from sqlalchemy import URL, Engine, MetaData, create_engine, inspect, text
+from sqlalchemy.engine import CursorResult
+from sqlalchemy.schema import CreateTable
+from sqlalchemy.types import NullType
+
+
+def get_mssql_table_names(engine: Engine):
+    inspector = inspect(engine)
+    schemas = inspector.get_schema_names()
+    db_tables: list[str] = []
+    for schema in schemas:
+        for table_name in inspector.get_table_names(schema=schema):
+            db_tables.append(f"{schema}.{table_name}")
+    return db_tables
+
+
+class DatalineSQLDatabase(SQLDatabase):
+
+    def __init__(
+        self,
+        engine: Engine,
+        schema: str | None = None,
+        metadata: MetaData | None = None,
+        ignore_tables: list[str] | None = None,
+        include_tables: list[str] | None = None,
+        sample_rows_in_table_info: int = 3,
+        indexes_in_table_info: bool = False,
+        custom_table_info: dict | None = None,
+        view_support: bool = True,  # default False in base class
+        max_string_length: int = 300,
+        lazy_table_reflection: bool = False,
+    ):
+        if engine.dialect.name == "mssql":
+            lazy_table_reflection = True  # so that we do our own reflection with multiple schemas
+        super().__init__(
+            engine,
+            schema,
+            metadata,
+            ignore_tables,
+            include_tables,
+            sample_rows_in_table_info,
+            indexes_in_table_info,
+            custom_table_info,
+            view_support,
+            max_string_length,
+            lazy_table_reflection,
+        )
+        if self.dialect == "mssql":
+            inspector = inspect(self._engine)
+            schemas = inspector.get_schema_names()
+            for curr_schema in schemas:
+                relevant_tables = [
+                    table.split(".")[1] for table in self._usable_tables if table.split(".")[0] == curr_schema
+                ]
+                if relevant_tables:
+                    self._metadata.reflect(
+                        views=view_support, bind=self._engine, only=relevant_tables, schema=curr_schema
+                    )
+
+    def custom_run_sql(self, query: str):
+        if self.dialect == "mssql":
+            with self._engine.begin() as connection:
+                command = text(query)
+                result = connection.execute(command)
+                rows = result.fetchall()
+                columns = list(result.keys())
+                return columns, rows
+
+        result = cast(CursorResult[Any], super().run(query, "cursor", include_columns=True))  # type: ignore[misc]
+        rows = result.fetchall()
+        columns = list(result.keys())
+        return columns, rows
+
+    @classmethod
+    def from_uri(cls, database_uri: str | URL, engine_args: dict | None = None, **kwargs: Any) -> Self:
```

Comment:
Why did yo uneed to redefine this one? Can't see any mssql specific stuff

Reply:
That's right, I only did this because the parent class method specifies a return type of `SQLDatabase`, and that's annoying because it creates type issues (benign). See screenshot for example
![image](https://github.com/user-attachments/assets/312e14b8-ceb5-43ed-9613-2b85e08916ba)